src/gui/src/histogramWidget.cpp

```diff
@@ -0,0 +1,242 @@
+{
+  chart_->setTitle("Endpoint Slack");
+
+  STAGuiInterface sta_gui(sta_);
+  
+  auto end_points = sta_gui.getEndPoints();
+  auto pin_iterator = end_points.begin();
+  std::vector<float> all_slack;
+  int unconstrained_count = 0;
+
+  for(int i = 0; i < end_points.size(); ++i) {
+    const int converter = 1e9;
+    float pin_slack = 0;
+    pin_slack = sta_gui.getPinSlack(*pin_iterator);
+    if(pin_slack != sta::INF)
+      all_slack.push_back(pin_slack*converter);
+    else
+      unconstrained_count++;
+    pin_iterator++;
+  }
+
+  if(unconstrained_count != 0) {
+    const QString label_message = "Number of unconstrained pins: ";
+    QString unconstrained_number;
+    unconstrained_number.setNum(unconstrained_count);
+    label_->setText(label_message + unconstrained_number);
+  }
+
+  auto max_slack = std::max_element(all_slack.begin(),
+                                    all_slack.end());
+
+  auto min_slack = std::min_element(all_slack.begin(),
+                                    all_slack.end());
+
+  //+1 for values around zero
+  int total_pos_buckets = *max_slack + 1;
+  int total_neg_buckets = abs(*min_slack) + 1;
+  int offset = abs(*min_slack);    
+  std::vector<float> pos_buckets[total_pos_buckets];
+  std::vector<float> neg_buckets[total_neg_buckets];
+
+  for(int i = 0; i < all_slack.size(); ++i) {
+    if(all_slack[i] < 0) {
+      int bucket_index = all_slack[i];
+      neg_buckets[bucket_index + offset].push_back(all_slack[i]);
+    } else {
+      int bucket_index = all_slack[i];
+      pos_buckets[bucket_index].push_back(all_slack[i]);
+    }
+  }
+
+  QBarSet* neg_set = new QBarSet("");
+  neg_set->setBorderColor(0x8b0000);  //darkred
+  neg_set->setColor(0xf08080);        //lightcoral
+
+  QBarSet* pos_set = new QBarSet("");
+  pos_set->setBorderColor(0x006400);  //darkgreen
+  pos_set->setColor(0x90ee90);        //lightgreen
+
+  QStringList time_values;
+  
+  const QString open_bracket = "[";
+  const QString close_parenthesis = ")";
+  const QString comma = ", ";
+  QString curr_value;
+  int curr_slack = 0;
+  int bucket_count = 0;
+  int max_y = 0;
+
+  for(int i = 0; i < total_neg_buckets; ++i) {
+    bucket_count = neg_buckets[i].size();
+    *neg_set << bucket_count;
+    *pos_set << 0;
+    curr_slack = neg_buckets[i][0];
+    QString prev_value;
+    time_values << open_bracket + prev_value.setNum(curr_slack-1) + comma + curr_value.setNum(curr_slack) + close_parenthesis;
+    if(max_y < bucket_count)
+      max_y = bucket_count;
+  }
+
+  for(int i = 0; i < total_pos_buckets; ++i) {
+    bucket_count = pos_buckets[i].size();
+    *pos_set << pos_buckets[i].size();
+    *neg_set << 0;
+    curr_slack = pos_buckets[i][0];
+    QString next_value;
+    time_values << open_bracket + curr_value.setNum(curr_slack) + comma + next_value.setNum(curr_slack+1) + close_parenthesis;
+    if(max_y < bucket_count)
+      max_y = bucket_count;
+  }
+
+  axis_x_->setTitleText("Slack [ns]");
```

Comment:
The time units depend on the .lib and should be taken from sta.  Some libraries may be in ps.